pkg/disk/partition_table.go

```diff
@@ -160,6 +161,291 @@ func NewPartitionTable(basePT *PartitionTable, mountpoints []blueprint.Filesyste
+	// PartitionTableType must be either "dos" or "gpt". Defaults to "gpt".
+	PartitionTableType string
+
+	// BootMode determines the types of boot-related partitions that are
+	// automatically added, BIOS boot (legacy), ESP (UEFI), or both (hybrid).
+	// If none, no boot-related partitions are created.
+	BootMode platform.BootMode
+
+	// DefaultFSType determines the filesystem type for automatically created
+	// filesystems and custom mountpoints that don't specify a type.
+	// None is only valid if no partitions are created and all mountpoints
+	// partitions specify a type.
+	DefaultFSType FSType
+
+	// RequiredSizes defines a map of minimum sizes for specific directories.
+	// These indirectly control the minimum sizes of partitions. A directory
+	// with a required size will set the minimum size of the partition with the
+	// mountpoint that contains the directory. Additional directory
+	// requirements are additive, meaning the minimum size for a mountpoint's
+	// partition is the sum of all the required directory sizes it will
+	// contain. By default, if no requiredSizes are provided, the new partition
+	// table will require at least 1 GiB for '/' and 2 GiB for '/usr'. In most
+	// cases, this translates to a requirement of 3 GiB for the root partition,
+	// Logical Volume, or Btrfs subvolume.
+	RequiredSizes map[string]uint64
+}
+
+// NewCustomPartitionTable creates a partition table based almost entirely on the partitioning customizations from a blueprint.
+func NewCustomPartitionTable(customizations *blueprint.PartitioningCustomization, options *CustomPartitionTableOptions, rng *rand.Rand) (*PartitionTable, error) {
+	pt := &PartitionTable{}
+
+	if options == nil {
+		// init options with empty struct to use defaults
+		options = &CustomPartitionTableOptions{
+			PartitionTableType: "gpt",
+		}
+	}
+
+	switch options.PartitionTableType {
+	case "gpt", "dos":
+		pt.Type = options.PartitionTableType
+	case "":
+		// default to "gpt"
+		pt.Type = "gpt"
+	default:
+		return nil, fmt.Errorf("invalid partition table type specified when generating partition table: %s", options.PartitionTableType)
+	}
+
+	switch options.BootMode {
+	case platform.BOOT_LEGACY:
+		// add BIOS boot partition
+		pt.Partitions = append(pt.Partitions, mkBIOSBoot())
+	case platform.BOOT_UEFI:
+		// add ESP
+		pt.Partitions = append(pt.Partitions, mkESP(200*common.MebiByte))
+	case platform.BOOT_HYBRID:
+		// add both
+		pt.Partitions = append(pt.Partitions, mkBIOSBoot())
+		pt.Partitions = append(pt.Partitions, mkESP(200*common.MebiByte))
+	case platform.BOOT_NONE:
+	default:
+		return nil, fmt.Errorf("invalid boot mode specified when generating partition table: %s", options.BootMode)
+	}
+
+	// The boot type will be the default only if it's a supported filesystem
+	// type for /boot (ext4 or xfs). Otherwise, we default to xfs.
+	var bootType FSType
+	switch options.DefaultFSType {
+	case FS_EXT4, FS_XFS:
+		bootType = options.DefaultFSType
+	case FS_NONE:
```

Comment:
Actually, NONE should be allowed.  I think we should make it so that `options.DefaultFSType == FS_NONE` is only an issue when a non-boot partition is automatically created.